### PR TITLE
[WIP][SPARK-35055][SQL] Fix the requirement in BitSet.union

### PR DESCRIPTION
**Pull request by sigmod**
### What changes were proposed in this pull request?

Reversed the requirement in BitSet.union:

require(this.numWords <= other.numWords)
=>
require(this.numWords >= other.numWords)

### Why are the changes needed?

Ensure that we don't go out-of-bound in a BitSet's Long array.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing tests.

### CODE REVIEW

**Review by c21 — COMMENTED** [review]
Are we sure this is the right fix? It seems that we loop over `this.numWords`, so `this.numWords <= other.numWords` seems right?

**Comment by sigmod** [comment]
> Are we sure this is the right fix? It seems that we loop over `this.numWords`, so `this.numWords <= other.numWords` seems right?

Yeah, you're right. 
FYI: @cloud-fan

File: core/src/main/scala/org/apache/spark/util/collection/BitSet.scala [hunk 1]
```diff
@@ -240,7 +240,7 @@ class BitSet(numBits: Int) extends Serializable {
    * Compute bit-wise union with another BitSet and overwrite bits in this BitSet with the result.
    */
   def union(other: BitSet): Unit = {
-    require(this.numWords <= other.numWords)
+    require(this.numWords >= other.numWords)
     var ind = 0
     while( ind < this.numWords ) {
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/util/collection/BitSet.scala:245]: hmm, so the `union` function here ignores the bits in `other` if it's longer than `this`?

Comment (sigmod) [reply core/src/main/scala/org/apache/spark/util/collection/BitSet.scala:245]: Yes. 
If we want prevent occasional bugs, we can change 
 require(this.numWords <= other.numWords)
to 
 require(this.numWords == other.numWords)?

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/util/collection/BitSet.scala:245]: yea, that looks better

Comment (sigmod) [reply core/src/main/scala/org/apache/spark/util/collection/BitSet.scala:245]: All right, will send a PR for that.

Comment (c21) [reply core/src/main/scala/org/apache/spark/util/collection/BitSet.scala:245]: Force equal length of two bitsets sounds fine. btw Java `java.util.BitSet` allows two bitsets to be different lengths (`or()`, `and()`, etc).

**Pull request closed without merging**